tokio/src/sync/semaphore_ll.rs

```diff
@@ -205,64 +238,82 @@ impl Semaphore {
                         }
 
-                        maybe_strong = Some(WaiterNode::into_non_null(waiter.clone()));
+                        maybe_waiter = Some(waiter);
                     } else {
-                        // If no `waiter`, then the task is not registered and there
-                        // is no further work to do.
+                        // No waiter, this indicates the caller does not wish to
+                        // "wait", so there is nothing left to do.
                         return Pending;
                     }
                 }
 
-                next.set_waiter(maybe_strong.unwrap());
+                next.set_waiter(maybe_waiter.unwrap());
             }
 
             debug_assert_ne!(curr.0, 0);
             debug_assert_ne!(next.0, 0);
 
-            match next.compare_exchange(&self.state, curr, AcqRel, Acquire) {
+            match self.state.compare_exchange(curr.0, next.0, AcqRel, Acquire) {
                 Ok(_) => {
-                    match curr.waiter() {
-                        Some(prev_waiter) => {
-                            let waiter = maybe_strong.unwrap();
-
-                            // Finish pushing
-                            unsafe {
-                                prev_waiter.as_ref().next.store(waiter.as_ptr(), Release);
-                            }
-
-                            return Pending;
+                    if acquired {
+                        // Successfully acquire permits **without** queuing the
+                        // waiter node. The waiter node is not currently in the
+                        // queue.
+                        revert_to_idle!();
+                        return Ready(Ok(()));
+                    } else {
+                        // The node is pushed into the queue, the final step is
+                        // to set the node's "next" pointer to return the wait
+                        // queue into a consistent state.
+
+                        let prev_waiter =
+                            curr.waiter().unwrap_or_else(|| NonNull::from(&*self.stub));
+
+                        let waiter = maybe_waiter.unwrap();
+
+                        // Link the nodes.
+                        //
+                        // Safety: the mpsc algorithm guarantees the old tail of
+                        // the queue is not removed from the queue during the
+                        // push process.
+                        unsafe {
+                            prev_waiter.as_ref().store_next(waiter);
                         }
-                        None => {
-                            undo_strong!();
 
-                            return Ready(Ok(()));
-                        }
+                        return Pending;
                     }
                 }
                 Err(actual) => {
-                    curr = actual;
+                    curr = SemState(actual);
                 }
             }
         }
```

Comment:
The acquire path is unrelated to the early return. They are working on separate atomic vals.

The `rx_lock` field limits a single thread to enter the critical section. Any thread that doesn't acquire the lock can return early because the work that has to happen is tracked by `rx_lock`. The thread that acquired the lock is forced to perform this work before releasing the lock.